packages/jest-cli/src/search_source.js

```diff
@@ -238,4 +239,37 @@ export default class SearchSource {
       return Promise.resolve({tests: []});
     }
   }
+
+  async getTestPaths(
+    globalConfig: GlobalConfig,
+    changedFilesPromise: ?ChangedFilesPromise,
+  ): Promise<SearchResult> {
+    const searchResult = await this._getTestPaths(
+      globalConfig,
+      changedFilesPromise,
+    );
+
+    const filterPath = globalConfig.filter;
+
+    if (filterPath) {
+      const tests = searchResult.tests;
+
+      // $FlowFixMe: dynamic require.
+      const filter = require(filterPath);
+      const filtered = await filter(tests.map(test => test.path));
+
+      if (!Array.isArray(filtered)) {
+        throw new Error(
+          `Filter ${filterPath} did not return a valid test list`,
+        );
+      }
+
+      // $FlowFixMe: Object.assign with empty object causes troubles to Flow.
+      return Object.assign({}, searchResult, {
+        tests: tests.filter(test => filtered.includes(test.path)),
```

Comment:
That way we avoid the O(n*m) algorithm here.

Reply:
What I can do for that is create a `Set`, then filter by doing a `.has`.